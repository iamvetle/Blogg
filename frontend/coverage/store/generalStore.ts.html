
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for store/generalStore.ts</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../index.html">All files</a> / <a href="index.html">store</a> generalStore.ts</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">95.23% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>40/42</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">100% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>1/1</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">0% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>0/1</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">95.23% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>40/42</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line high'></div>
    <pre><table class="coverage">
<tr><td class="line-count quiet"><a name='L1'></a><a href='#L1'>1</a>
<a name='L2'></a><a href='#L2'>2</a>
<a name='L3'></a><a href='#L3'>3</a>
<a name='L4'></a><a href='#L4'>4</a>
<a name='L5'></a><a href='#L5'>5</a>
<a name='L6'></a><a href='#L6'>6</a>
<a name='L7'></a><a href='#L7'>7</a>
<a name='L8'></a><a href='#L8'>8</a>
<a name='L9'></a><a href='#L9'>9</a>
<a name='L10'></a><a href='#L10'>10</a>
<a name='L11'></a><a href='#L11'>11</a>
<a name='L12'></a><a href='#L12'>12</a>
<a name='L13'></a><a href='#L13'>13</a>
<a name='L14'></a><a href='#L14'>14</a>
<a name='L15'></a><a href='#L15'>15</a>
<a name='L16'></a><a href='#L16'>16</a>
<a name='L17'></a><a href='#L17'>17</a>
<a name='L18'></a><a href='#L18'>18</a>
<a name='L19'></a><a href='#L19'>19</a>
<a name='L20'></a><a href='#L20'>20</a>
<a name='L21'></a><a href='#L21'>21</a>
<a name='L22'></a><a href='#L22'>22</a>
<a name='L23'></a><a href='#L23'>23</a>
<a name='L24'></a><a href='#L24'>24</a>
<a name='L25'></a><a href='#L25'>25</a>
<a name='L26'></a><a href='#L26'>26</a>
<a name='L27'></a><a href='#L27'>27</a>
<a name='L28'></a><a href='#L28'>28</a>
<a name='L29'></a><a href='#L29'>29</a>
<a name='L30'></a><a href='#L30'>30</a>
<a name='L31'></a><a href='#L31'>31</a>
<a name='L32'></a><a href='#L32'>32</a>
<a name='L33'></a><a href='#L33'>33</a>
<a name='L34'></a><a href='#L34'>34</a>
<a name='L35'></a><a href='#L35'>35</a>
<a name='L36'></a><a href='#L36'>36</a>
<a name='L37'></a><a href='#L37'>37</a>
<a name='L38'></a><a href='#L38'>38</a>
<a name='L39'></a><a href='#L39'>39</a>
<a name='L40'></a><a href='#L40'>40</a>
<a name='L41'></a><a href='#L41'>41</a>
<a name='L42'></a><a href='#L42'>42</a>
<a name='L43'></a><a href='#L43'>43</a></td><td class="line-coverage quiet"><span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import { defineStore } from "pinia";
import { PersonalPostType } from '../typescript/interfaces';
&nbsp;
export const useGeneralStore = defineStore("general", () =&gt; {
&nbsp;
  // Local storage
  const isAuthenticated = ref(false);
  const username = ref&lt;string | null&gt;(null);
&nbsp;
  // Posts
  const posts = ref&lt;ArticlesSnippetsType | null&gt;(null)
&nbsp;
  // Min side
  const theUser = ref&lt;PersonalPostType | null&gt;(null) 
  const personalPosts = ref&lt;ArticlesSnippetsType | null&gt;(null)
&nbsp;
  // Links
  const baseFeedURL = "http://localhost:8888/api/feed/"
  const post_snippets_url = ref&lt;string&gt;("http://localhost:8888/api/feed/")
  const personal_post_snippets_url = ref&lt;string&gt;("http://localhost:8888/api/min-side/posts/")
&nbsp;
  const next_page_link = ref&lt;string | null&gt;(null)
  const previous_page_link = ref&lt;string | null&gt;(null)
  const last_page_link = ref&lt;string | null&gt;(null)
&nbsp;
  // pagnation
  const total_pages_count = ref&lt;number | null&gt;(null);
  const number_of_posts_count = ref&lt;number | null&gt;(null);
  const current_page = ref&lt;number | null&gt;(null);
&nbsp;
  // error pages
  const apiDownError = ref(false);
&nbsp;
  //other
  const search_bar_show = ref(false)
&nbsp;
  const changeAuthenticated = <span class="fstat-no" title="function not covered" >(state: boolean) =&gt; {</span>
<span class="cstat-no" title="statement not covered" >    isAuthenticated.value = state;</span>
<span class="cstat-no" title="statement not covered" >  };</span>
&nbsp;
  return { posts, theUser, search_bar_show, personalPosts, personal_post_snippets_url, current_page, baseFeedURL, post_snippets_url, isAuthenticated, apiDownError, username, total_pages_count, number_of_posts_count, next_page_link, previous_page_link, last_page_link, changeAuthenticated };
});
&nbsp;</pre></td></tr></table></pre>

                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2023-10-03T20:43:48.495Z
            </div>
        <script src="../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../sorter.js"></script>
        <script src="../block-navigation.js"></script>
    </body>
</html>
    