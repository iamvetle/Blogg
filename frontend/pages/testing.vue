<template>
  <div class="flex">

    <form @submit.prevent method="post" class="flex flex-col" enctype="multipart/form-data">

      <label for="name">Name:</label>
      <input type="text" name="name" id="name">

      <label for="email">Email</label>
      <input type="email" name="email" id="email">

      <label for="message">Message:</label>
      <textarea name="message" id="" cols="30" rows="10" placeholder="Write a message here"></textarea>

      <label for="file">File:</label>
      <input type="file" name="file" id="file">

      <button type="submit">Submit</button>
    </form>

  </div>
</template>

<script setup lang="ts">
onMounted(() => {

  const form = document.querySelector("form")

  form?.addEventListener("submit", (e: any) => {


    const formData = new FormData(form)

    for (let item of formData) {
      console.log(item[0], item[1])
    }
    
    // ta axios of prÃ¸v fetch her
  })





})

// watchEffect(() => {

// })

</script>

<style scoped></style>