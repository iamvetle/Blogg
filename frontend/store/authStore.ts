import { defineStore } from 'pinia';

/**
 * @fileoverview
 * This file defines a Pinia store for managing authentication-related state in a Nuxt 3 application.
 */

export const useAuthStore = defineStore('Auth Store for managing authentication-related state', () => {

    /** Stores the authentication token. */
    const token = ref('');

    /** Stores the username of the logged-in user. */
    const username = ref('');

    /**
     * Computed property reflecting the current authentication status.
     * @returns - The authentication status.
     * 
     * This property is automatically updated based on the presence of the authentication token.
     */
    const isAuthenticated = computed(() => {
        return Boolean(token.value);
    });

    /** 
     * Watch for changes in localStorage and update the token accordingly.
     */
    watchEffect(() => {
        if (process.client) {
            const tokenFromStorage = localStorage.getItem('token');

            // Update token if present in localStorage
            token.value = tokenFromStorage ?? '';
        }

    });

    /** 
     * Watch for changes in localStorage and update the username accordingly.
     */
    watchEffect(() => {
        if (process.client) {
            const usernameFromStorage = localStorage.getItem('username');

            // Update username if present in localStorage
            username.value = usernameFromStorage ?? '';
        }

    });

    /**
     * Sets the authentication token in localStorage.
     * @param tokenInput - The authentication token.
     */
    const setTokenToLocalStorage = (tokenInput: string): void => {
        const tokenValue = tokenInput ?? '';
        localStorage.setItem('token', tokenValue);
    }

    /**
     * Sets the username in localStorage.
     * @param usernameInput - The username.
     */
    const setUsernameToLocalStorage = (usernameInput: string): void => {
        const usernameValue = usernameInput ?? '';
        localStorage.setItem('username', usernameValue);
    }

    /** 
     * Removes the authentication token from localStorage and sets the authentication state to false.
     */
    const removeTokenFromLocalStorage = (): void => {
        localStorage.removeItem('token');
    }

    /** 
     * Removes the username from localStorage.
     */
    const removeUsernameFromLocalStorage = (): void => {
        localStorage.removeItem('username');
    }

    /** 
     * Retrieves the authentication token.
     * @returns - The authentication token or null if empty.
     */
    const retrieveToken = (): string | null => {
        return token.value === '' ? null : token.value;
    }

    /** 
     * Sets up the initial state based on the localStorage token on component mount.
     */
    const authStoreSetup = (): void => {
        console.log('authStoreSetup was actually called');
        const tokenFromLocalStorage = localStorage.getItem('token');
        console.log(tokenFromLocalStorage);

        // Assign an empty string, "", if there was no localStorage token
        token.value = tokenFromLocalStorage ?? '';
    };

    /** 
     * Returns the computed property and the reactive variable for external use.
     */
    return {
        isAuthenticated, token, username,
        setUsernameToLocalStorage, setTokenToLocalStorage,
        removeTokenFromLocalStorage, removeUsernameFromLocalStorage,
        retrieveToken, authStoreSetup
    };
});
