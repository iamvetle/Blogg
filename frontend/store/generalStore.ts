import { defineStore } from "pinia";

export const useGeneralStore = defineStore("general", () => {

  // Local storage
  const isAuthenticated = ref(false);
  const username = ref<string | null>(null);

  // Posts
  const posts = ref<SnippetPostMultipleType | null>(null)

  // [id] user
  const theUser = ref<NormalUserProfileType | null>(null) 

  // Min side
  const personalPosts = ref<SnippetPostMultipleType | null>(null)
  const personalUser = ref<LoggedInUserProfileType | null>(null)

  // Links
  const baseFeedURL = "http://localhost:8888/api/search/"
  const post_snippets_url = ref<string>("http://localhost:8888/api/feed/")
  const personal_post_snippets_url = ref<string>("http://localhost:8888/api/minkonto/posts/")

  const next_page_link = ref<string>("")
  const previous_page_link = ref<string>("")
  const last_page_link = ref<string>("")

  // pagnation
  const total_pages_count = ref<number>(0);
  const number_of_posts_count = ref<number>(0);
  const current_page = ref<number>(1);

  // error pages
  const apiDownError = ref(false);

  //other
  const search_bar_show = ref(false)

  /** Contains names of all the tags possible */
  const allTags = ref<CategoryType[]>([])

  const idArrayOfSavedPosts = ref<number[]>([])

  const changeAuthenticated = (state: boolean) => {
    isAuthenticated.value = state;
  };

  return { posts, allTags, idArrayOfSavedPosts, personalUser, theUser, search_bar_show, personalPosts, personal_post_snippets_url, current_page, baseFeedURL, post_snippets_url, isAuthenticated, apiDownError, username, total_pages_count, number_of_posts_count, next_page_link, previous_page_link, last_page_link, changeAuthenticated };
});
