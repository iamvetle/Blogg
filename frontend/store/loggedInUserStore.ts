import { defineStore } from 'pinia';

export const useLoggedInUserStore = defineStore("Store for storing information about the logged in user", () => {
    
    /**
     * Contains all of the information about the logged-in user
     */
    const loggedInUserProfile = ref<LoggedInUserProfileType | null>(null) // personalUser


    /**
     * The id of all of the posts the logged in user has saved is stored here
     */
    const idArrayOfSavedPosts = ref<number[]>([])

    /**
     * Keeps the username of all the users the logged-in user is following
     */
    const idArrayOfLoggedInUserFollowingUsers = ref<string[]>([])



    return { loggedInUserProfile, idArrayOfSavedPosts, idArrayOfLoggedInUserFollowingUsers }
})

