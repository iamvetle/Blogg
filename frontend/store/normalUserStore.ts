export const useNormalUserStore = defineStore("Store for storing user information", () => {
    
    /**
     * All information about a user of 'user/[id].vue' will be here.
     */

    const normalUserProfile = ref<NormalUserProfileType | null>(null) // theUser


    return { normalUserProfile }
})