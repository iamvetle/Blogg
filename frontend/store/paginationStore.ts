import { defineStore } from 'pinia';

export const usePaginationStore = defineStore("Pagination Store", () => {
    /**
     * This store contains links associated pagination.
     * 
     * It is used to switch between pages. They also updated by search and initial load.
     * 
     */

    /**
     * This url gets update frequently. It is supposed to represent the 
     * last url a fetch call for index posts was made to.
     */

    const postStore = usePostStore()

    const activeFetchURL = ref<any>(urls.api.posts.feed)
    
    const next_page = ref<null | string>(null)
    const previous_page = ref<null | string>(null) 

    const countOfPosts = ref<number>(0);

    const currentPageNumber = ref<number>(1);

    const setPagination = () => {
        if (postStore.posts != null) {
    
            countOfPosts.value = postStore.posts.count
            next_page.value = postStore.posts.next
            previous_page.value = postStore.posts.previous
        }
    }
    
    const route = useRoute()
    const router = useRouter()

    /** Adds the current page as a query to the local route url */
    watchEffect(() => {
        router.replace({
            query: {
                ...route.query,
                page:currentPageNumber.value
            }
        })        
    })


    /**
     * resets everything
     */
    const resetStore = () => {
        activeFetchURL.value = urls.api.posts.feed
        next_page.value = ""
        previous_page.value = ""
    
        countOfPosts.value = 0
    
        currentPageNumber.value = 0 
    }

    return { activeFetchURL, resetStore, setPagination, next_page, previous_page, countOfPosts, currentPageNumber }
})