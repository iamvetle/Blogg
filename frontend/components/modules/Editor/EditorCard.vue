<template>
	<div class="p-2" v-if="editor">
		<div id="editor-container"
			class="w-full px-[60px] pt-[35px] pb-[30px] bg-white flex flex-col text-gray-800 rounded-lg min-h-[450px] mb-12"
			@click="editor.commands.focus()">

			<div v-if="showModal">
				<teleport to="#modal">
					<div class="w-full h-screen blur-sm">
						<Modal @confirm-published="publishPost" @cancel-published="cancelPublishing" />
					</div>
				</teleport>
			</div>

			<div class="w-full not-prose mb-6">
				<EditorFloatingMenu :editor="editor" @addImage="addImage" @setLink="setLink"
					@toggleHeading1="toggleHeading(1)" @toggleHeading2="toggleHeading(2)"
					@setHorizontalRule="horizontalRule" />

				<!-- <EditorBubbleMenu
				:editor="editor"
				/> -->
				<EditorCardTopMenu :editor="editor"/>
			</div>

			<hr class="not-prose mb-8">

			<div data-test="direct-editor" class="w-full">
				<editor-content :editor="editor" />
			</div>

		</div>

		<hr class="mb-4">

		<div data-test="cancel_publish_buttons" class="buttons flex">

			<BaseButton id="cancel"
				class="btn border border-secondary-low p-1 px-4 font-semibold cursor-pointer text-gray-500 hover:text-gray-400 hover:border-secondary-base ml-auto"
				@click="buttonCancelClick" text="Cancel" />
			<BaseButton id="publish"
				class="btn border border-indigo-base p-1 px-4 font-semibold cursor-pointer text-plain ml-2 bg-secondary-base hover:bg-secondary-low"
				@click="buttonTryPublishClick" text="Publish" />

		</div>
	</div>
</template>

<script setup lang="ts">

import { useEditor, EditorContent } from '@tiptap/vue-3'
// import { BubbleMenu } from '@tiptap/vue-3';
import Document from '@tiptap/extension-document' // required
import BulletList from '@tiptap/extension-bullet-list'
import CodeBlock from '@tiptap/extension-code-block'
import HardBreak from '@tiptap/extension-hard-break'
import HorizontalRule from '@tiptap/extension-horizontal-rule'
import ListItem from '@tiptap/extension-list-item'
import OrderedList from '@tiptap/extension-ordered-list'
import Code from '@tiptap/extension-code'
// import Strike from '@tiptap/extension-strike'
import Paragraph from '@tiptap/extension-paragraph'
import Text from '@tiptap/extension-text'
import Heading from '@tiptap/extension-heading'
import Italic from '@tiptap/extension-italic'
import Link from '@tiptap/extension-link'
import Bold from '@tiptap/extension-bold'
import Blockquote from '@tiptap/extension-blockquote'
import Dropcursor from '@tiptap/extension-dropcursor'
import Gapcursor from '@tiptap/extension-gapcursor'
import History from '@tiptap/extension-history'
import Underline from '@tiptap/extension-underline'
import Image from '@tiptap/extension-image'

import { useGeneralStore } from '~/store/generalStore';

const generalStore = useGeneralStore()
const emit = defineEmits(['newPostMaterial'])

const props = defineProps<{
	initialPost?: string;
}>();


// const errorHappened = ref(null)

/**
 * This variable dictates whether the Modal is shown or not.
 * 
 * True: Modal is shown. 
 * False: Modal is hidden.
*/
const showModal = ref(false)
const html = ref<string | null | undefined>(null);
const route = useRoute()

/** This two are just so that I can share them between the button 
 * click function and the modal publish function*/
const title = ref<string | null | undefined>(null);
const body = ref<string | null | undefined>(null);

const editor: any = useEditor({ //@ts-ignore
	"type": "doc",
	content: '',
	extensions: [
		Heading.configure({
			levels: [1, 2, 3]
		}),
		Blockquote,
		// Subscript,
		// Placeholder.configure({
		// 	placeholder: '123',
		// 	showOnlyWhenEditable: false,
		// 	includeChildren: true,	
		// 	showOnlyCurrent: false
		// }),
		BulletList,
		Image.configure({
			allowBase64: true,
		}),
		CodeBlock.configure({
			exitOnTripleEnter: true,
			languageClassPrefix: 'javascript',
		}),
		HardBreak.configure({
			keepMarks: false,
		}),
		HorizontalRule,
		ListItem,
		OrderedList,
		// Superscript,
		// Subscript,
		// Strike,
		Gapcursor,
		// Table.configure({
		// 	resizable: true,
		// }),
		// TableRow,
		// TableHeader,
		// TableCell,
		History,
		// Youtube.configure({
		// nocookie: true,
		// }),
		Dropcursor,
		Document,
		Paragraph,
		Text, // required
		// TaskList,
		// TaskItem.configure({
		// nested: true,
		// }),
		Italic,
		Link.configure({ //@ts-ignore
			validate: href => /^https?:\/\//.test(href),
		}),
		Bold,
		Underline,
		Code,
		// Youtube,
	],

	editorProps: {
		attributes: {
			class: 'focus:outline-none',
			// class: 'prose prose-sm sm:prose lg:prose-lg xl:prose-2xl mx-auto focus:outline-none',	
		},
	},
	autofocus: true
})

const unwatch = watch(() => props.initialPost, (newPost) => {
	if (newPost) {
		editor.value.commands.setContent(newPost);
		unwatch(); // Unwatch after the first trigger
	}
}, {
	immediate: false // This will ensure the watcher doesn't trigger on initial setup
});

/**
 * Tracks the "emptyness" of the editor
 */
// const isEditorEmpty = computed(() => !editor.value?.content?.trim());

/**
 * On update it takes and updates the HTML value?
 */
onMounted(() => {
	editor.value?.on("update", () => {
		html.value = editor.value?.getHTML();
	});
});

// BUTTON ACTIONS

/**
 * Tries to publish the post. First, extracts the title and content and then
 * shows a modal for confirmation.
 */
const buttonTryPublishClick = async () => {
	html.value = editor.value?.getHTML();
	const answer = extractTitleAndContent(html.value as string);

	title.value = answer?.title;
	body.value = answer?.body;

	if (!title.value || !body.value) {
		alert("Invalid post content");
		return;
	}
	showModal.value = true;
	generalStore.turnBackgroundForModel("blur-sm");
};

/**
 * Cancels the post creation and navigates to the feed(home) page.
 */
const buttonCancelClick = () => {
	const router = useRouter();
	// Clears the input
	editor.value?.commands.clearContent();
	router.push('/');
};

// MODAL EMITS/EVENTS 2/2

/** The modal can return TWO things */

// 1/2

/** 
 * Button action
 * 
 * Emits the new post data and clears the editor content.
 * 
 * * Only called by modal
 */
const publishPost = () => {
	showModal.value = false;
	generalStore.turnBackgroundForModel(null);
	emit('newPostMaterial', { title: title.value, content: body.value });
	editor.value?.chain().focus().clearContent().run();
};

// 2/2
/**
 * Cancels publishing and resets the modal state.
 */
const cancelPublishing = () => {
	showModal.value = false;
	generalStore.turnBackgroundForModel(null);
};


/** METHODS FOR THE EDITOR */

const setLink = () => {
	const previousUrl = editor.value.getAttributes('link').href
	const url = window.prompt('URL', previousUrl)

	// cancelled
	if (url === null) {
		return
	}

	// empty
	if (url === '') {
		editor.value
			.chain()
			.focus()
			.extendMarkRange('link')
			.unsetLink()
			.run()

		return
	}

	// update link
	editor.value
		.chain()
		.focus()
		.extendMarkRange('link')
		.setLink({ href: url })
		.run()
}

function addImage() {
	const url = window.prompt('URL')

	if (url) {
		editor.value.chain().focus().setImage({ src: url }).run()
	}
}

const toggleHeading = (level: number) => {
	editor.value.chain().focus().toggleHeading({ level }).run();
};

const horizontalRule = () => {
	editor.value.chain().focus().setHorizontalRule().run()
};

// const toggleUnderline = () => {
// 	editor.value.chain().focus().toggleUnderline().run()
// }

// const toggleCode = () => {
// 	editor.value.chain().focus().toggleCode().run()
// }

// const toggleBlockquote = () => {
// 	editor.value.chain().focus().toggleBlockquote().run()

// }

// const toggleBold = () => {
// 	editor.value.chain().focus().toggleBold().run()
// }

// const toggleItalic = () => {
// 	editor.value.chain().focus().toggleItalic().run()
// }


/**
 * Fils the content if there is a post content inside of sessionstorage
 */
onMounted(() => {
	if (editor) {

		// If the route is /edit/ it wont retrieve cached halway done post

		if (route.path.includes("edit") === false) {
			const htmlPost = sessionStorage.getItem("htmlPost")

			if (htmlPost) {
				editor.value.chain().focus().insertContent(htmlPost).run()
			}
		}
	}
})

/**
 * This saves the content into sessionstorage
 */
onUnmounted(() => {
	/** 
 * Retrieves the text that has been written in the 
 * text editor, as HTML.
 */
	const htmlPost = editor.value?.getHTML()

	sessionStorage.setItem("htmlPost", htmlPost)

})





</script>

<style scoped>
.editor-placeholder {
	color: grey;
	font-style: italic;
	/* Additional styling as needed */
}
</style>