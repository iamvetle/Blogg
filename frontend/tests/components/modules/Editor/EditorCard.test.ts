import { VueWrapper, mount } from '@vue/test-utils'
import { createTestingPinia } from '@pinia/testing'
import { useGeneralStore } from '~/store/generalStore'
import EditorCard from '~/components/modules/Editor/EditorCard.vue'

describe('_blank testing', () => {
    let wrapper:VueWrapper
    let store;
    let pinia;

    beforeEach(() => {
        pinia = createTestingPinia()
        store = useGeneralStore(pinia)
        wrapper = mount(EditorCard, {
            global: {
                plugins: [pinia],
            },
        })
    })

    test('exists', () => {
        expect(wrapper.exists()).toBe(true)
    })
})