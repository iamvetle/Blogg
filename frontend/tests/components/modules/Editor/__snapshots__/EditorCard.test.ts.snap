// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing the editorcard (very small bad tests becasue of a loff rewriting) > Should match snapshot 1`] = `
"<div class=\\"p-2\\">
  <div id=\\"editor-container\\" class=\\"w-full px-[60px] pt-[35px] pb-[30px] flex flex-col rounded-lg min-h-[450px] mb-24\\">
    <div data-test=\\"text_editor_modals\\">
      <!-- The Modal to confirm to the post -->
      <!--v-if-->
      <!-- Using Nuxt UI, don't need teleport -->
      <!--v-if-->
      <!-- The Modal to discard the content post -->
      <!--v-if-->
    </div><!-- Menus for editor -->
    <!--v-if-->
    <div data-test=\\"editor_and_title_input\\" class=\\"mt-2 w-full mx-auto\\">
      <!-- Input editor (not prose)-->
      <div class=\\"max-w-2xl mx-auto\\">
        <inputtext placeholder=\\"Title\\" modelvalue=\\"\\" modelmodifiers=\\"[object Object]\\" class=\\"not-prose mx-auto pb-3 w-full border-none bg-inherit text-4xl leading-4 font-extrabold outline-none placeholder:text-gray-300\\"></inputtext>
      </div>
      <hr class=\\"not-prose invisible\\"><!-- Main Editor (has prose) -->
      <div data-test=\\"direct-editor\\" class=\\"max-w-2xl mx-auto pt-3 w-full min-h-screen prose px-1\\">
        <editor-content-stub></editor-content-stub>
      </div>
    </div>
  </div>
  <hr class=\\"mb-4\\">
</div>"
`;
